nomad/server.go

```diff
@@ -1171,6 +1166,13 @@ func (s *Server) setupRpcServer(server *rpc.Server, ctx *RPCContext) {
 		s.staticEndpoints.Namespace = &Namespace{srv: s}
 		s.staticEndpoints.Enterprise = NewEnterpriseEndpoints(s)
 
+		// Endpoints that need the RPC context are added but not registered
+		s.staticEndpoints.Alloc = &Alloc{srv: s, logger: s.logger.Named("alloc")}
+		s.staticEndpoints.Deployment = &Deployment{srv: s, logger: s.logger.Named("deployment")}
+		s.staticEndpoints.Eval = &Eval{srv: s, logger: s.logger.Named("eval")}
+		s.staticEndpoints.Node = &Node{srv: s, logger: s.logger.Named("client")}
+		s.staticEndpoints.Plan = &Plan{srv: s, logger: s.logger.Named("plan")}
+
```

Comment:
Hm, I'm not sure this is required for any endpoints other than `Node` and the compiler should help you check.

I think we only register the Node endpoint on staticEndpoints so the heartbeater can call methods on it directly without actually having to go over RPC. I don't think that applies to any of the other methods.

If possible let's remove the other fields from the `endpoints` struct entirely and see if the compiler complains. Putting dynamic endpoints on the staticEndpoints structs is definitely a bit of a smell.

Reply:
> Would it be worth having a separate filled, like dynamicEndpoints?

Probably not as even the "dynamic" endpoints have to be called statically when called directly instead of over RPC, so "static" isn't really misleading in that case...

...this poses an interesting problem for `ctx` because it will be `nil` in these cases, so panics not only seem easy to accidentally cause, but they'll be on the leader!

Making `ctx` an interface with `Local` and `RPC` implementations is probably optimal, but not necessary for this PR. I think I pointed out the only place where a nil ctx could panic, but it would be worth quick checking.

Another option would be refactoring the dynamic-RPCs-called-statically to have their actual logic in a separate func that can be called directly. The RPC handler method can then just be a light wrapper to do RPC-specific operations like auth.


Again, nothing is a blocker here, but some refactoring would make writing correct code easier here.

Reply:
Yeah, that makes sense. The `static` name confused me a bit, but I think I get it now 🙂

I also ran into the empty `ctx` problem until I realized where to set it properly. I added the `nil` check as you mentioned in the other comment 👍